python/lsst/butler/units.py

```diff
@@ -0,0 +1,992 @@
+# 
+# LSST Data Management System
+#
+# Copyright 2008-2017  AURA/LSST.
+# 
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+# 
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+# 
+# You should have received a copy of the LSST License Statement and 
+# the GNU General Public License along with this program.  If not, 
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+
+from collections import OrderedDict
+from hashlib import sha512
+from datetime import datetime  # placeholder while prototyping
+from sqlalchemy.sql import select, and_, exists
+from .schema import CameraTable, AbstractFilterTable, PhysicalFilterTable, PhysicalSensorTable, VisitTable, ObservedSensorTable, SnapTable, VisitRangeTable, SkyMapTable, TractTable, PatchTable
+
+
+class DataUnitMeta(type):
+    """
+    Metaclass for DataUnit, keeps track of all subclasses to enable construction by name.
+    """
+    def __init__(cls, name, bases, dct):
+        if not hasattr(cls, '_subclasses'):
+            # this is the base class.
+            cls._subclasses = {}
+        else:
+            cls._subclasses[name] = cls
+
+        super(DataUnitMeta, cls).__init__(name, bases, dct)
+
+
+class DataUnit(metaclass=DataUnitMeta):
+
+    def insert(self, connection):
+        raise NotImplementedError("pure virtual")
+
+    def pkey(self):
+        return NotImplemented("pure virtual")
+
+    def invariantHash(self):
+        """
+        Compute a hash that is invariant accross Python sessions, and hence can be stored in a database.
+        """
+        return sha512(b''.join(str(v).encode() for v in self.pkey)).digest()
```

Comment:
If you want to be invariant I suggest you specify the encoding rather than relying on the default.